File: labs/mathlogic/lab3/lab3_3.py
from typing import Optional, List

class AVLTreeNode:
    def __init__(self, data: int):
        self.data: int = data
        self.height: int = 1
        self.left: Optional['AVLTreeNode'] = None
        self.right: Optional['AVLTreeNode'] = None

class AVLTree:
    def __init__(self):
        self.root: Optional[AVLTreeNode] = None
        self._size: int = 0

    def _is_empty(self):
        return self._size == 0

    def _search(self, key: int, node: AVLTreeNode) -> AVLTreeNode:
        if key < node.data:
            if node.left is None:
                return node
            return self._search(key, node.left)
        if key > node.data:
            if node.right is None:
                return node
            return self._search(key, node.right)
        return node

    def _direction_search(self, node: AVLTreeNode, direction: str) -> AVLTreeNode:
        if direction == "left":
            if node.left is None:
                return node
            return self._direction_search(node.left, direction)
        elif direction == "right":
            if node.right is None:
                return node
            return self._direction_search(node.right, direction)
        return node

    def _inorder(self, node: Optional[AVLTreeNode], result: List[int]):
        if node is None:
            return
        self._inorder(node.left, result)
        result.append(node.data)
        self._inorder(node.right, result)

    def _insert_node(self, key: int, node: Optional[AVLTreeNode]) -> AVLTreeNode:
        if node is None:
            self._size += 1
            return AVLTreeNode(key)

        if key < node.data:
            node.left = self._insert_node(key, node.left)
        if key > node.data:
            node.right = self._insert_node(key, node.right)
        if key == node.data:
            return node

        node.height = 1 + max(self.height(node.left), self.height(node.right))
        balance = self.balance_factor(node)

        if balance > 1 and key < node.left.data:
            return self.rotate_right(node)
        
        if balance < -1 and key > node.right.data:
            return self.rotate_left(node)
        
        if balance > 1 and key > node.left.data:
            node.left = self.rotate_left(node.left)
            return self.rotate_right(node)
        
        if balance < -1 and key < node.right.data:
            node.right = self.rotate_right(node.right)
            return self.rotate_left(node)

        return node

    def _delete_node(self, key: int, node: Optional[AVLTreeNode]) -> Optional[AVLTreeNode]:
        if node is None:
            return None

        if key < node.data:
            node.left = self._delete_node(key, node.left)
        elif key > node.data:
            node.right = self._delete_node(key, node.right)
        elif key == node.data:
            if node.left is None and node.right is None:
                self._size -= 1
                return None
            if node.left is None and node.right is not None:
                self._size -= 1
                return node.right
            if node.left is not None and node.right is None:
                self._size -= 1
                return node.left
            
            temp = self._direction_search(node.right, "left")
            node.data = temp.data
            node.right = self._delete_node(temp.data, node.right)

        node.height = 1 + max(self.height(node.left), self.height(node.right))
        balance = self.balance_factor(node)

        if balance > 1 and self.balance_factor(node.left) >= 0:
            return self.rotate_right(node)
        
        if balance > 1 and self.balance_factor(node.left) < 0:
            node.left = self.rotate_left(node.left)
            return self.rotate_right(node)
        
        if balance < -1 and self.balance_factor(node.right) <= 0:
            return self.rotate_left(node)
        
        if balance < -1 and self.balance_factor(node.right) > 0:
            node.right = self.rotate_right(node.right)
            return self.rotate_left(node)

        return node

    def height(self, node: Optional[AVLTreeNode]) -> int:
        if node is None:
            return 0
        return node.height

    def balance_factor(self, node: Optional[AVLTreeNode]) -> int:
        if node is None:
            return 0
        return self.height(node.left) - self.height(node.right)

    def rotate_right(self, node: AVLTreeNode) -> AVLTreeNode:
        left_child = node.left
        right_of_left_child = left_child.right

        left_child.right = node
        node.left = right_of_left_child

        node.height = 1 + max(self.height(node.left), self.height(node.right))
        left_child.height = 1 + max(self.height(left_child.left), self.height(left_child.right))

        return left_child

    def rotate_left(self, node: AVLTreeNode) -> AVLTreeNode:
        right_child = node.right
        left_of_right_child = right_child.left

        right_child.left = node
        node.right = left_of_right_child

        node.height = 1 + max(self.height(node.left), self.height(node.right))
        right_child.height = 1 + max(self.height(right_child.left), self.height(right_child.right))

        return right_child

    def insert(self, key: int):
        self.root = self._insert_node(key, self.root)

    def delete(self, key: int):
        if self._is_empty():
            raise Exception("Tree is empty!")
        self.root = self._delete_node(key, self.root)

    def find(self, key: int) -> Optional[AVLTreeNode]:
        if self._is_empty():
            return None
        node = self._search(key, self.root)
        if node.data != key:
            return None
        return node

    def traverse_inorder(self):
        result: List[int] = []
        if self._is_empty():
            return result
        self._inorder(self.root, result)
        return result

File: labs/mathlogic/lab3/lab3_3.ts
export class AVLTreeNode {
  data: number;
  height: number = 1;
  left: AVLTreeNode | null = null;
  right: AVLTreeNode | null = null;

  constructor(data: number) {
    this.data = data;
  }
}

export class AVLTree {
  root: AVLTreeNode | null = null;
  private _size: number = 0;

  private _is_empty() {
    return this._size == 0;
  }

  private _search(key: number, node: AVLTreeNode): AVLTreeNode {
    if (key < node.data) {
      if (!node.left) {
        return node;
      }
      return this._search(key, node.left);
    }
    if (key > node.data) {
      if (!node.right) {
        return node;
      }
      return this._search(key, node.right);
    }
    return node;
  }

  private _directionSearch(node: AVLTreeNode, direction: "left" | "right") {
    switch (direction) {
      case "left":
        if (!node.left) return node;
        return this._directionSearch(node.left, direction);
      case "right":
        if (!node.right) return node;
        return this._directionSearch(node.right, direction);
    }
  }

  private _inorder(node: AVLTreeNode | null, result: number[]) {
    if (!node) return;
    this._inorder(node.left, result);
    result.push(node.data);
    this._inorder(node.right, result);
  }

  private _insertNode(key: number, node: AVLTreeNode | null): AVLTreeNode {
    if (!node) {
      this._size++;
      return new AVLTreeNode(key);
    }

    if (key < node.data) {
      node.left = this._insertNode(key, node.left);
    }
    if (key > node.data) {
      node.right = this._insertNode(key, node.right);
    }
    if (key == node.data) {
      return node;
    }

    node.height = 1 + Math.max(this.height(node.left), this.height(node.right));
    const balance = this.balance_factor(node);

    if (balance > 1 && key < node.left!.data) {
      return this.rotate_right(node);
    }
    if (balance < -1 && key > node.right!.data) {
      return this.rotate_left(node);
    }
    if (balance > 1 && key > node.left!.data) {
      node.left = this.rotate_left(node.left!);
      return this.rotate_right(node);
    }
    if (balance < -1 && key < node.right!.data) {
      node.right = this.rotate_right(node.right!);
      return this.rotate_left(node);
    }

    return node;
  }

  private _deleteNode(
    key: number,
    node: AVLTreeNode | null,
  ): AVLTreeNode | null {
    if (!node) return null;

    if (key < node.data) {
      node.left = this._deleteNode(key, node.left);
    }
    if (key > node.data) {
      node.right = this._deleteNode(key, node.right);
    }
    if (key == node.data) {
      if (!node.left && !node.right) {
        this._size--;
        return null;
      }
      if (!node.left && !!node.right) {
        this._size--;
        return node.right;
      }
      if (!!node.left && !node.right) {
        this._size--;
        return node.left;
      }
      const temp = this._directionSearch(node.right, "left");
      node.data = temp.data;
      node.right = this._deleteNode(temp.data, node.right);
    }

    node.height = 1 + Math.max(this.height(node.left), this.height(node.right));
    const balance = this.balance_factor(node);

    if (balance > 1 && this.balance_factor(node.left) >= 0) {
      return this.rotate_right(node);
    }
    if (balance > 1 && this.balance_factor(node.left) < 0) {
      node.left = this.rotate_left(node.left!);
      return this.rotate_right(node);
    }
    if (balance < -1 && this.balance_factor(node.right) <= 0) {
      return this.rotate_left(node);
    }
    if (balance < -1 && this.balance_factor(node.right) > 0) {
      node.right = this.rotate_right(node.right!);
      return this.rotate_left(node);
    }

    return node;
  }

  height(node: AVLTreeNode | null): number {
    if (!node) return 0;
    return node.height;
  }

  balance_factor(node: AVLTreeNode | null): number {
    if (!node) return 0;
    return this.height(node.left) - this.height(node.right);
  }

  rotate_right(node: AVLTreeNode): AVLTreeNode {
    const leftChild = node.left!;
    const rightOfLeftChild = leftChild.right;

    leftChild.right = node;
    node.left = rightOfLeftChild;

    node.height = 1 + Math.max(this.height(node.left), this.height(node.right));
    leftChild.height =
      1 + Math.max(this.height(leftChild.left), this.height(leftChild.right));

    return leftChild;
  }

  rotate_left(node: AVLTreeNode): AVLTreeNode {
    const rightChild = node.right!;
    const leftOfRightChild = rightChild.left;

    rightChild.left = node;
    node.right = leftOfRightChild;

    node.height = 1 + Math.max(this.height(node.left), this.height(node.right));
    rightChild.height =
      1 + Math.max(this.height(rightChild.left), this.height(rightChild.right));

    return rightChild;
  }

  insert(key: number) {
    this.root = this._insertNode(key, this.root);
  }

  delete(key: number) {
    if (this._is_empty()) {
      throw new Error("Tree is empty!");
    }
    this.root = this._deleteNode(key, this.root);
  }

  find(key: number): AVLTreeNode | null {
    if (this._is_empty()) {
      return null;
    }
    const node = this._search(key, this.root!);
    if (node.data != key) return null;
    return node;
  }

  traverse_inorder() {
    const result: number[] = [];
    if (this._is_empty()) {
      return result;
    }
    this._inorder(this.root, result);
    return result;
  }
}
